### PR TITLE
Build documentation with Sphinx

### LINKED ISSUE #22: Build documentation with Sphinx
Documentation in the README and obejct APIs (already mostly in a Sphinx-ready format) should be documented with Sphinx.

- Rename current `docs` directory to `openapi`. Reflect this name change in the README and/or new .rst docs.
- Port README to .rst
- Add new documentation to `docs` directory
- Make formatting changes as necessary to docstrings if the existing format results in something not human-readable. For example a reference to `porter.somemodule.SomeClass` that doesn't result in a hyper link is okay since it's still clear where to find the object. If hyperlinks can be easily supported we can tackle that now, otherwise we'll get the basic docs together and add the hyperlinks at another time.
- *Edit*: Update the Makefile so that `make openapi` generates the openapi docs (this is currently `make docs`) and `make docs` generates the sphinx docs.
- *Edit*: Update the `pre-commit-hook` to generate both openapi and sphinx docs.

### PR DESCRIPTION
**Pull request by cadent-mrichman**
This pull requests provides Read The Docs themed package reference and basic user manual based on the pre-existing README.  Resolves #22, including updating docstrings for hyperlinking cross references.

### CODE REVIEW

**Comment by cadent-mrichman** [comment]
@cadent-dgates

File: docs/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+_build
```
Comment (cadent-dgates) [inline docs/.gitignore]: @cadent-mrichman is this file necessary? Or can it be removed? Seems like it could be moved to the root-level .gitignore

File: docs/conf.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+#
+# import os
+# import sys
+# sys.path.insert(0, os.path.abspath('.'))
+
+
+# -- Project information -----------------------------------------------------
+
+project = 'porter'
+copyright = '2020, Cadent'
```
Comment (cadent-dgates) [inline docs/conf.py]: @cadent-mrichman, good thinking adding the copyright here. Let's double check internally exactly how this should be specified. Same for the author set immediately below.

File: docs/make.bat [hunk 3]
```diff
@@ -0,0 +1,35 @@
+@ECHO OFF
```
Comment (cadent-dgates) [inline docs/make.bat]: @cadent-mrichman is this autogenerated by Sphinx?

File: docs/rest_api.rst [hunk 4]
```diff
@@ -0,0 +1,30 @@
+.. _rest_api:
+
+REST API
+========
+
+
+The complete details of an API exposed by ``porter`` can be found in the `openapi <https://openapi.tools/>`_ spec in `this repository <https://github.com/cadent-mrichman/porter/tree/u/mrichman/docs/templates>`_.
```
Comment (cadent-dgates) [inline docs/rest_api.rst]: @cadent-mrichman this links to your personal repository. Instead we should use a "permalink" to CadentTech.

File: docs/rest_api.rst [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+Additionally you can open the `static documentation <https://github.com/cadent-mrichman/porter/blob/u/mrichman/docs/templates/html/index.html>`_ generated from this spec in your web browser. On a Mac:
```
Comment (cadent-dgates) [inline docs/rest_api.rst]: @cadent-mrichman this links to your personal repository. Instead we should use a "permalink" to CadentTech.

File: porter/config.py [hunk 6]
```diff
@@ -1,3 +1,5 @@
+"""Configuration options for debugging, not recommended for use in production apps."""
```
Comment (cadent-dgates) [inline porter/config.py]: @cadent-mrichman Can we change this to just "Configuration options." and leave the comments below. Setting return_request_id to True or False is fine for production depending on user preference. It's just the traceback and user data errors that should be used with caution as the comment indicates.

File: docs/.gitignore (hunk 1 above)
Comment (cadent-mrichman) [reply docs/.gitignore]: I agree that this file can be removed.

File: docs/conf.py (hunk 2 above)
Comment (cadent-mrichman) [reply docs/conf.py]: This is something `sphinx-quickstart` requests in the initial setup.  I agree that we should make a deliberate choice here rather than just a guess.

**Comment by cadent-dgates** [comment]
@cadent-mrichman One last thing: if we can get this hosted on Read the Docs in a timely fashion, we should include the link at the top of the README before merging this PR

File: docs/make.bat (hunk 3 above)
Comment (cadent-mrichman) [reply docs/make.bat]: Yes, make.bat and Makefile are both autogenerated.

File: docs/rest_api.rst (hunk 4 above)
Comment (cadent-mrichman) [reply docs/rest_api.rst]: Good point.  It was handy to be able to test this, but I'll set it up to link to the file that will exist once this is merged.

File: porter/config.py (hunk 6 above)
Comment (cadent-mrichman) [reply porter/config.py]: Gotcha – will change.

**Comment by cadent-mrichman** [comment]
@cadent-dgates Ok, most of the above are addressed in the last few commits.

I don't think we can build this on ReadTheDocs until this PR is merged.  So I'm leaning towards waiting for a second PR to trim the README (reducing duplication) and link to the docs once they are hosted.

I think the only other outstanding issue above is authorship.

**Comment by cadent-mrichman** [comment]
@cadent-dgates Added note on AWS tests in docs/tests.rst; let me know if you think this is fine or if it belongs elsewhere.

**Comment by cadent-mrichman** [comment]
@cadent-dgates Just update author and copyright to credit "Cadent Data Science" per offline discussion.

**Comment by cadent-dgates** [comment]
@cadent-mrichman looks good, two things

1. The LICENSE states "Copyright (c) 2019 CadentTech". Whichever way we go this should be the same value we are adding to the Sphinx docs.
2. The note on running tests looks good, can we add that the environment variables are needed to test porter's loading capabilities with a link to the file?

**Comment by cadent-dgates** [comment]
@cadent-mrichman The docstrings in `datascience.py` in this PR conflict with the changes we just merged in that added the WrappedModel validations.

Is there anything else that needs to be addressed before merging this, or is this PR otherwise good to go?
